hashes/zkevm/src/keccak/computation_circuit/circuit.rs

```diff
@@ -0,0 +1,343 @@
+    type Config = KeccakComputationConfig<F>;
+    type FloorPlanner = SimpleFloorPlanner;
+    type Params = KeccakComputationCircuitParams;
+
+    fn params(&self) -> Self::Params {
+        self.config_params.clone()
+    }
+
+    /// Creates a new instance of the [RangeCircuitBuilder] without witnesses by setting the witness_gen_only flag to false
+    fn without_witnesses(&self) -> Self {
+        unimplemented!()
+    }
+
+    /// Configures a new circuit using [`BaseConfigParams`]
+    fn configure_with_params(meta: &mut ConstraintSystem<F>, params: Self::Params) -> Self::Config {
+        let range_circuit_config = RangeCircuitBuilder::configure(meta);
+        let keccak_circuit_config = KeccakCircuitConfig::new(meta, params.keccak_circuit_params);
+        Self::Config { range_circuit_config, keccak_circuit_config }
+    }
+
+    fn configure(_: &mut ConstraintSystem<F>) -> Self::Config {
+        unreachable!("You must use configure_with_params");
+    }
+
+    fn synthesize(
+        &self,
+        config: Self::Config,
+        mut layouter: impl Layouter<F>,
+    ) -> Result<(), Error> {
+        let k = self.config_params.k;
+        config.keccak_circuit_config.load_aux_tables(&mut layouter, k as u32)?;
+        // TODO: do I need SKIP_FIRST_PASS?
+        let (keccak_rows, _) = multi_keccak::<F>(
+            &self.inputs,
+            Some(self.capacity),
+            self.params().keccak_circuit_params,
+        );
+        let mut keccak_assigned_rows: Vec<KeccakAssignedRow<'_, F>> = Vec::default();
+        layouter.assign_region(
+            || "keccak circuit",
+            |mut region| {
+                keccak_assigned_rows =
+                    config.keccak_circuit_config.assign(&mut region, &keccak_rows);
+                Ok(())
+            },
+        )?;
+        let loaded_assigned_rows = self.load_keccak_assigned_rows(keccak_assigned_rows);
+
+        let (compact_inputs, result_selector_per_chunk) =
+            self.generate_poseidon_inputs(&loaded_assigned_rows);
+
+        let _circuit_final_output = self.compute_circuit_final_output(
+            &loaded_assigned_rows,
+            &compact_inputs,
+            &result_selector_per_chunk,
+        );
+
+        self.range_circuit_builder.borrow().synthesize(config.range_circuit_config, layouter)?;
+        Ok(())
+    }
+}
+
+/// Witnesses to be exposed as circuit outputs.
+struct KeccakCircuitOutput<F: Field> {
+    pub(crate) input_poseidon: AssignedValue<F>,
+    pub(crate) hash_lo: AssignedValue<F>,
+    pub(crate) hash_hi: AssignedValue<F>,
+}
+
+struct LoadedKeccakAssignedRow<F: Field> {
+    pub(crate) is_final: AssignedValue<F>,
+    pub(crate) hash_lo: AssignedValue<F>,
+    pub(crate) hash_hi: AssignedValue<F>,
+    pub(crate) bytes_left: AssignedValue<F>,
+    pub(crate) word_value: AssignedValue<F>,
+}
+
+impl<F: Field> KeccakComputationCircuit<F> {
+    /// Load keccak assigned rows into halo2-lib.
+    fn load_keccak_assigned_rows(
```

Comment:
Do you need every row to be loaded? I feel it's some unnecessary overhead - you could 
1. skip dummy rows
2. use fact that everything is in clumps of rows_per_round and only read first row from each row_per_round